client/components/calypso-i18n-provider/index.tsx

```diff
@@ -2,21 +2,23 @@
  * External dependencies
  */
 import * as React from 'react';
-import { setLocaleData as setWpI18nLocaleData } from '@wordpress/i18n';
-import { I18nProvider } from '@automattic/react-i18n';
+import { createI18n, setLocaleData as setWpI18nLocaleData } from '@wordpress/i18n';
+import { I18nProvider } from '@wordpress/react-i18n';
 import i18n from 'i18n-calypso';
 import { LocaleProvider, i18nDefaultLocaleSlug } from '@automattic/i18n-utils';
 
 const CalypsoI18nProvider: React.FunctionComponent = ( { children } ) => {
 	const [ localeData, setLocaleData ] = React.useState( i18n.getLocale() );
 	const [ localeSlug, setLocaleSlug ] = React.useState( i18n.getLocaleSlug() );
+	const wpI18n = React.useMemo( () => createI18n( localeData ), [ localeSlug ] );
 
 	React.useEffect( () => {
 		const onChange = () => {
 			// Create a new object to ensure the provider sees the data has changed
 			const nextLocaleData = Object.assign( {}, i18n.getLocale() );
 
 			setWpI18nLocaleData( nextLocaleData );
+			wpI18n.setLocaleData( nextLocaleData );
```

Comment:
This component could be radically simplified, as it currently stores `localeData` at three places and we need only one.

Creating the `wpI18n` instance should not be needed. The locale data are already in the default `i18n` instance and we are setting them with `setWpI18nLocaleData`.

The `localeData` local state can also be removed. It doesn't seem to be used by anything else bug `createI18n( localeData )`. That create call can be removed and even if it couldn't, we could call `createI18n( i18n.getLocale() )`.

Reply:
We do need to created new `@wordpress/i18n` instance on locale change, because even though the name of the `setLocaleData` method suggests that it will set the locale data, what it actually does it to [merge the passed locale data with the existing one](https://github.com/WordPress/gutenberg/blob/0eb76dcbdb22ef198c690ade215acb56604edd00/packages/i18n/src/create-i18n.js#L163-L176). This means that if we decide to use the same instance across locale changes, it may end up displaying a translations from previous locales for translations entries that are missing in the current locale data, instead of falling back to showing the original (English) string. This behaviour would be observed in case the user switches from Mag-16 to a non-Mag-16 language, as the latter usually have incomplete translations.

The way `@automattic/react-i18n` currently works is that it would create an i18n instance when the locale data changes, but I believe that the reason why it was implemented like that was because `setLocaleData` was not available back then. Currently, it seems to be redundant, which is why I've decided to memoize it just by the locale slug.

As for the locale data state - you're right, it's no longer needed, so I've removed it.

Reply:
In that case, why do we use the default `i18n` instance and call `setLocaleData` on it? That way, the default instance will contain a mixture of merged translations after a few locale switches.

Reply:
I've created a PR with the proposed `setLocaleData` enhancement here - https://github.com/WordPress/gutenberg/pull/30419.